### ISSUE TITLE
GPU: OpenXR integration

### ISSUE THREAD

**Issue by Beyley** [libsdl-org/SDL#10925]
I'm not entirely sure on how this could be made in a way that doesn't force SDL_gpu to just flat out expose device/framebuffer handles, but I think it could be possible without compromising that aspect of the API.

I think the main things that would be required are:
Method which returns the name of the OpenXR graphics device extension required for a specified `SDL_GPUDevice *`, eg. `"XR_KHR_D3D11_enable"`. Allowing the app developer to pass that name directly into their instance creation, without directly needing to handle each and every backend themselves.
```c
// psuedo-code
const char *SDL_GetGPUXREnableExtensionName(SDL_GPUDevice* device)
{
	switch(device->backend)
	{
		case D3D11:
			return XR_KHR_D3D11_ENABLE_EXTENSION_NAME;
		case VULKAN:
			return ...;
		// etc.
	}
}
```

Method which calls `xrCreateSession` for you, filling in the `createInfo->next` pointer with the graphics binding info (eg. a `XrGraphicsBindingD3D11KHR` struct), filling that graphics binding struct's `next` pointer with the `next` pointer inside `createInfo->next`, then calling `xrCreateSession` for you.
eg.
```c
// psuedo-code
XrResult SDL_CreateGPUXRSession(SDL_GPUDevice* device, XrInstance instance, const XrSessionCreateInfo *createInfo, XrSession *session)
{
	XrSessionCreateInfo newCreateInfo = *createInfo;

	switch(device->backend)
	{
		case D3D11:
			XrGraphicsBindingD3D11KHR binding = {
				.type = XR_TYPE_GRAPHICS_BINDING_D3D11_KHR,
				.next = createInfo->next,
				.device = device->backend_d3d11->d3d11_device,
			};

			newCreateInfo.next = &binding;
			break;
		case VULKAN:
			// etc etc.
			break;
	}

	return xrCreateSession(instance, &newCreateInfo, session);
}
```
This prevents the internal device from being leaked into the application.

A similar thing can be done with the swapchain creation functions, wrapping any OpenXR APIs which require access to the texture handles to prevent them from leaking into the application.

Not sure how feasable it is for this specific method of integrating it, but I think having some way of cleanly integrating OpenXR into SDL_gpu is an important usecase that should be looked into.

**Comment by flibitijibibo** [comment]
Agreed - in the nuclear case we can just add a property but it'd be nice to have a full list of what apps need to support VR with any given GPU API.

We could maybe kill two birds with one stone by having an OpenXR sample at https://github.com/TheSpydog/SDL_gpu_examples, and developing a gpu-xr PR to support it. My only experience with XR right now is input, for #4464, so I dunno how much help I'll be.

**Comment by Beyley** [comment]
> it'd be nice to have a full list of what apps need to support VR with any given GPU API.

I've updated the issue with a full breakdown of the init process for every graphics API, turns out this will probably be a lot more involved than I originally thought. But should still be possible without exposing any raw GPU handles to the end user.

**Comment by Beyley** [comment]
I have started *very* early work on this [here](https://github.com/Beyley/SDL) and [here](https://github.com/Beyley/SDL_gpu_examples). Right now I have instance/system/session creation working on Vulkan with `KHR_vulkan_enable2`.

I have nothing visual to show yet, since I still gotta write the rest of the surrounding code in the example itself, and also write some swapchain related functions in SDL_gpu to let me render to the XR session swapchain. 

Only after that can I actually start working on cleaning this up to upstream.

**Comment by corentin35000** [comment]
Is this still relevant today? I would be very interested in developing VR games next!

**Comment by Beyley** [comment]
> Is this still relevant today? I would be very interested in developing VR games next!

I haven't worked on my fork in about a month or so, but I'm still interested in making this happen, I'm just very busy with other projects at the moment, feel free to contribute to my fork if you want to see this make progress. Right now it desperately needs a rebase

**Comment by Beyley** [comment]
After some hacking today, i managed to get us all the way to rendering to the swapchain (with some hacks/hardcoded values)

![image](https://github.com/user-attachments/assets/4b35d939-5935-4c11-8448-5b3dd1a37a5a)
![image](https://github.com/user-attachments/assets/1136e2a0-e00d-455e-bb71-41cec519b0cd)

big problem is that Monado's debug view is showing massive texture corruption (however the main simulated HMD's view is displaying correctly!)
im assuming this is some UB im hitting with vulkan, but given the app crashes in renderdoc (and the fact I don't really know much vulkan), I'm not _entirely_ sure yet how to debug it, but i've at least made some progress!

**Comment by Beyley** [comment]
another update: i got it all working!
top left is a "desktop view of the scene", rendered myself to an SDL window using SDL_gpu
middle is Monado's debug interface, showing the composited image raw
bottom right is the GUI used for controlling the QWERTY virtual HMD, showing a readback from the GPU of the composited image (should always match the debug interface's view)
![image](https://github.com/user-attachments/assets/0990bee2-c69c-42ed-bdf2-951fb27c6453)

not drawing anything *pretty*, but it's all working, aside from some vulkan validation issues on app close im not _entirely_ sure how to solve yet.

**Comment by Starkium** [comment]
Is this still the best available option?

**Comment by Beyley** [comment]
> Is this still the best available option?

Could you elaborate what your question means? OpenXR is the current most-supported standard for cross-platform and cross-vendor VR and AR experiences, with multiple compliant runtimes available on all platforms (except macOS, where there's only a single simulator runtime [Meta XR Simulator]).

**Comment by Starkium** [comment]
> > Is this still the best available option?
> 
> Could you elaborate what your question means? OpenXR is the current most-supported standard for cross-platform and cross-vendor VR and AR experiences, with multiple compliant runtimes available on all platforms (except macOS, where there's only a single simulator runtime [Meta XR Simulator]).

Is this still the best way to work with openxr and SDL? was really hoping I didn't need to make numerous entry points for all different renderers and platforms. XR is the biggest offender. SDL supports OVR, or the other way around, but not openxr. 

Looking around for the "best"/simplest option for tying together cross platform support / mostly uniform engine path for stuff.

**Comment by Beyley** [comment]
> Is this still the best way to work with openxr and SDL? 

This issue just describes the _internal to SDL_ work required to wire up OpenXR with SDL GPU.
 
> SDL supports OVR, or the other way around, but not openxr

SDL has a video driver for OpenVR, but that's much different than what this issue is talking about, which is integration of OpenXR's graphical session initialization into SDL GPU to allow you to use SDL_gpu to render onto OpenXR swapchains.

> Looking around for the "best"/simplest option for tying together cross platform support / mostly uniform engine path for stuff.

If you're dead set on SDL GPU, I'd recommend taking the fork of SDL in #11601, rebasing it, and using that. It supports Vulkan fairly well, Direct3D 12 with some probably easy to fix validation errors, and as of now, no Metal support. It's also tested to work on both Windows and Linux

**Comment by Starkium** [comment]
I'll take a look at it, but I was considering using vulkan outside of sdl because of more features. Sorry, I didn't realize this was for sdl gpu api only.
